src/passes/OptimizeInstructions.cpp

```diff
@@ -1195,145 +1206,146 @@ struct OptimizeInstructions
-      if (binary->op == Abstract::getBinary(type, Abstract::Add) ||
-          binary->op == Abstract::getBinary(type, Abstract::Sub)) {
-        auto value = constRight;
-        if (value == 0x40 || value == 0x2000 || value == 0x100000 ||
-            value == 0x8000000 || value == 0x400000000LL ||
-            value == 0x20000000000LL || value == 0x1000000000000LL ||
-            value == 0x80000000000000LL || value == 0x4000000000000000LL) {
-          right->value = right->value.neg();
-          if (binary->op == Abstract::getBinary(type, Abstract::Add)) {
-            binary->op = Abstract::getBinary(type, Abstract::Sub);
-          } else {
-            binary->op = Abstract::getBinary(type, Abstract::Add);
-          }
-          return binary;
+      // of two 7 bits etc. higher. we therefore prefer x - -64 over x + 64. in
+      // theory we could just prefer negative numbers over positive, but that
+      // can have bad effects on gzip compression (as it would mean more
+      // subtractions than the more common additions). TODO: Simplify this by
+      // adding an ival matcher than can bind int64_t vars.
+      int64_t value;
+      if ((matches(curr, binary(Abstract::Add, any(), ival(&value))) ||
+           matches(curr, binary(Abstract::Sub, any(), ival(&value)))) &&
+          (value == 0x40 || value == 0x2000 || value == 0x100000 ||
+           value == 0x8000000 || value == 0x400000000LL ||
+           value == 0x20000000000LL || value == 0x1000000000000LL ||
+           value == 0x80000000000000LL || value == 0x4000000000000000LL)) {
+        right->value = right->value.neg();
+        if (matches(curr, binary(Abstract::Add, any(), constant()))) {
+          curr->op = Abstract::getBinary(type, Abstract::Sub);
+        } else {
+          curr->op = Abstract::getBinary(type, Abstract::Add);
         }
+        return curr;
       }
     }
-    if (type.isFloat()) {
-      auto value = right->value.getFloat();
-      if (value == 0.0) {
-        if (binary->op == Abstract::getBinary(type, Abstract::Sub)) {
-          if (std::signbit(value)) {
-            // x - (-0.0)   ==>   x + 0.0
-            binary->op = Abstract::getBinary(type, Abstract::Add);
-            right->value = right->value.neg();
-            return binary;
-          } else {
-            // x - 0.0   ==>   x
-            return binary->left;
-          }
-        } else if (binary->op == Abstract::getBinary(type, Abstract::Add)) {
-          if (std::signbit(value)) {
-            // x + (-0.0)   ==>   x
-            return binary->left;
-          }
+    {
+      double value;
+      if (matches(curr, binary(Abstract::Sub, any(), fval(&value)))) {
+        // x - (-0.0)   ==>   x + 0.0
+        if (std::signbit(value)) {
+          curr->op = Abstract::getBinary(type, Abstract::Add);
+          right->value = right->value.neg();
+          return curr;
+        } else {
+          // x - 0.0   ==>   x
+          return curr->left;
         }
```

Comment:
Hmm, it should check `value` with zero I guess:
```cpp
if (value == 0.0) {
  // x - (-0.0)   ==>   x + 0.0
  if (std::signbit(value)) {
    ...
  } else {
    ...
  }
}
```

Reply:
Or you could generalize `x - (-C)` ==> `x + C`. In this case:
```cpp
if (matches(curr, binary(Abstract::Sub, any(), fval(&value)))) {
  // x - (-C)   ==>   x + C
  if (std::signbit(value)) {
    curr->op = Abstract::getBinary(type, Abstract::Add);
    right->value = right->value.neg();
    return curr;
  } else {
    if (value == 0.0) {
      // x - 0.0   ==>   x
      return curr->left;
    }
  }
}
```

Reply:
Good catch, thanks! Is that transformation of `X - (-C)` always correct?

Reply:
Yes, even with non-canonical `-nan` witch become canonical `nan`. Hope fuzz tests also happy with this

Reply:
Even though it's small, let's leave that generalization to a follow-up PR. This PR is already large and complex, so it will be helpful to avoid functional changes in OptimizeInstructions.